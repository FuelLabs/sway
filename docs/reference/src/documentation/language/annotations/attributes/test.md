# Test

Sway provides the `#[test]` attribute which enables unit tests to be written in Sway.

## Success case

The `#[test]` attribute indicates that a test has passed if it did not revert.

```sway
{{#include ../../../../code/language/annotations/src/main.sw:success_test}}
```

## Revert Case

To test a case where code should revert we can use the `#[test(should_revert)]` annotation. If the test reverts then it will be reported as a passing test.

```sway
{{#include ../../../../code/language/annotations/src/main.sw:revert_test}}
```

We may specify a code to specifically test against.

```sway
{{#include ../../../../code/language/annotations/src/main.sw:revert_code_test}}
```
