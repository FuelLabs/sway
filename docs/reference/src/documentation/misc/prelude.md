# Standard Library Prelude

The [prelude](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/prelude.sw) is a list of commonly used features from the [standard library](https://github.com/FuelLabs/sway/tree/master/sway-lib-std) which is automatically imported into every Sway program.

The prelude contains the following:

- [`Address`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/address.sw): A struct containing a `b256` value which represents the wallet address
- [`ContractId`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/contract_id.sw) A struct containing a `b256` value which represents the ID of a contract
- [`Identity`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/identity.sw): An enum containing `Address` & `ContractID` structs
- [`Vec`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/vec.sw): A growable, heap-allocated vector
- [`StorageMap`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/storage.sw): A key-value mapping in contract storage
- [`Option`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/option.sw): An enum containing either some generic value `<T>` or an absence of that value, we also expose the variants directly:
  - `Some`
  - `None`
- [`Result`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/result.sw): An enum used to represent either a success or failure of an operation, we also expose the variants directly:
  - `Ok`
  - `Err`
- [`assert`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/assert.sw): A module containing
  - `assert`: A function that reverts the VM if the condition provided to it is false
  - `assert_eq`: A function that reverts the VM and logs its two inputs v1 and v2 if the condition v1 == v2 is false
  - `assert_ne`: A function that reverts the VM and logs its two inputs v1 and v2 if the condition v1 != v2 is false
- [`revert`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/revert.sw): A module containing
  - `require`: A function that reverts and logs a given value if the condition is `false`
  - `revert`: A function that reverts
- [`log`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/logging.sw): A function that logs arbitrary stack types
- [`msg_sender`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/auth.sw): A function that gets the Identity from which a call was made
