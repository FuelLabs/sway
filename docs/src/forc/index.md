# Forc

Forc stands for Fuel Orchestrator. Forc provides a variety of tools and commands for developers working with the Fuel ecosystem, such as scaffolding a new project, formatting, running scripts, deploying contracts, testing contracts, and more. If you're coming from a Rust background, forc is similar to cargo.

If you are new to Forc, see the [Forc Project](../introduction/forc_project.md) introduction section.

For a comprehensive overview of the Forc CLI commands, see the [Commands](./commands/index.md) section.
