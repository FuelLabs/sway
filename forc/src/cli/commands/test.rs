use crate::ops::forc_build;
use anyhow::Result;
use clap::Parser;
use std::io::{BufRead, BufReader};
use std::process;
use std::thread;

/// Run Rust-based tests on current project.
/// As of now, `forc test` is a simple wrapper on
/// `cargo test`; `forc init` also creates a rust
/// package under your project, named `tests`.
/// You can opt to either run these Rust tests by
/// using `forc test` or going inside the package
/// and using `cargo test`.
#[derive(Debug, Parser)]
pub(crate) struct Command {
    /// If specified, only run tests containing this string in their names
    pub test_name: Option<String>,
    /// Options passed through to the `cargo test` invocation.
    ///
    /// E.g. Given the following:
    ///
    /// `forc test --cargo-test-opts="--color always"`
    ///
    /// The `--color always` option is forwarded to `cargo test` like so:
    ///
    /// `cargo test --color always`
    #[clap(long)]
    pub cargo_test_opts: Option<String>,
    /// All trailing arguments following `--` are collected within this argument.
    ///
    /// E.g. Given the following:
    ///
    /// `forc test -- foo bar baz`
    ///
    /// The arguments `foo`, `bar` and `baz` are forwarded on to `cargo test` like so:
    ///
    /// `cargo test -- foo bar baz`
    #[clap(raw = true)]
    pub cargo_test_args: Vec<String>,
}

pub(crate) fn exec(command: Command) -> Result<()> {
    // Ensure the project builds before running tests.
    forc_build::build(Default::default())?;

    let mut cmd = process::Command::new("cargo");
    cmd.arg("test");

    // Pass through cargo test options.
    let mut user_specified_color_opt = false;
    if let Some(opts) = command.cargo_test_opts {
        user_specified_color_opt = opts.contains("--color");
        for opt in opts.split_whitespace() {
            cmd.arg(&opt);
        }
    }

    // If the coloring option wasn't specified by the user, enable it ourselves. This is useful as
    // `cargo test`'s coloring is disabled by default when run as a child process.
    if !user_specified_color_opt {
        cmd.args(&["--color", "always"]);
    }

    // Pass through test name.
    if let Some(ref name) = command.test_name {
        cmd.arg(name);
    }

    // Pass through cargo test args.
    if !command.cargo_test_args.is_empty() {
        cmd.arg("--");
        cmd.args(&command.cargo_test_args);
    }

    let mut child = cmd
        .stdout(process::Stdio::piped())
        .stderr(process::Stdio::piped())
        .spawn()
        .unwrap();

    let out = BufReader::new(child.stdout.take().unwrap());
    let err = BufReader::new(child.stderr.take().unwrap());

    // Reading stderr on a separate thread so we keep things non-blocking
    let thread = thread::spawn(move || {
        err.lines().for_each(|line| println!("{}", line.unwrap()));
    });

    out.lines().for_each(|line| println!("{}", line.unwrap()));
    thread.join().unwrap();

    Ok(())
}
