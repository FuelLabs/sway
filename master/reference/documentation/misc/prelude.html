<!DOCTYPE HTML>
<html lang="en" class="light sidebar-visible" dir="ltr">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>Standard Library Prelude - The Sway Reference</title>


        <!-- Custom HTML head -->

        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff">

        <link rel="icon" href="../../favicon.svg">
        <link rel="shortcut icon" href="../../favicon.png">
        <link rel="stylesheet" href="../../css/variables.css">
        <link rel="stylesheet" href="../../css/general.css">
        <link rel="stylesheet" href="../../css/chrome.css">
        <link rel="stylesheet" href="../../css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="../../FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="../../fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="../../highlight.css">
        <link rel="stylesheet" href="../../tomorrow-night.css">
        <link rel="stylesheet" href="../../ayu-highlight.css">

        <!-- Custom theme stylesheets -->


        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "../../";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>
        <!-- Start loading toc.js asap -->
        <script src="../../toc.js"></script>
    </head>
    <body>
    <div id="body-container">
        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            const html = document.documentElement;
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add("js");
        </script>

        <input type="checkbox" id="sidebar-toggle-anchor" class="hidden">

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var sidebar = null;
            var sidebar_toggle = document.getElementById("sidebar-toggle-anchor");
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            sidebar_toggle.checked = sidebar === 'visible';
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <!-- populated by js -->
            <mdbook-sidebar-scrollbox class="sidebar-scrollbox"></mdbook-sidebar-scrollbox>
            <noscript>
                <iframe class="sidebar-iframe-outer" src="../../toc.html"></iframe>
            </noscript>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle">
                <div class="sidebar-resize-indicator"></div>
            </div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky">
                    <div class="left-buttons">
                        <label id="sidebar-toggle" class="icon-button" for="sidebar-toggle-anchor" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </label>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">The Sway Reference</h1>

                    <div class="right-buttons">
                        <a href="../../print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>
                        <a href="https://github.com/FuelLabs/sway" title="Git repository" aria-label="Git repository">
                            <i id="git-repository-button" class="fa fa-github"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="standard-library-prelude"><a class="header" href="#standard-library-prelude">Standard Library Prelude</a></h1>
<p>The <a href="https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/prelude.sw">prelude</a> is a list of commonly used features from the <a href="https://github.com/FuelLabs/sway/tree/master/sway-lib-std">standard library</a> which is automatically imported into every Sway program.</p>
<p>The prelude contains the following:</p>
<ul>
<li><a href="https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/address.sw"><code>Address</code></a>: A struct containing a <code>b256</code> value which represents the wallet address</li>
<li><a href="https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/contract_id.sw"><code>ContractId</code></a> A struct containing a <code>b256</code> value which represents the ID of a contract</li>
<li><a href="https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/identity.sw"><code>Identity</code></a>: An enum containing <code>Address</code> &amp; <code>ContractID</code> structs</li>
<li><a href="https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/vec.sw"><code>Vec</code></a>: A growable, heap-allocated vector</li>
<li><a href="https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/storage.sw"><code>StorageMap</code></a>: A key-value mapping in contract storage</li>
<li><a href="https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/option.sw"><code>Option</code></a>: An enum containing either some generic value <code>&lt;T&gt;</code> or an absence of that value, we also expose the variants directly:
<ul>
<li><code>Some</code></li>
<li><code>None</code></li>
</ul>
</li>
<li><a href="https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/result.sw"><code>Result</code></a>: An enum used to represent either a success or failure of an operation, we also expose the variants directly:
<ul>
<li><code>Ok</code></li>
<li><code>Err</code></li>
</ul>
</li>
<li><a href="https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/assert.sw"><code>assert</code></a>: A module containing
<ul>
<li><code>assert</code>: A function that reverts the VM if the condition provided to it is false</li>
<li><code>assert_eq</code>: A function that reverts the VM and logs its two inputs v1 and v2 if the condition v1 == v2 is false</li>
<li><code>assert_ne</code>: A function that reverts the VM and logs its two inputs v1 and v2 if the condition v1 != v2 is false</li>
</ul>
</li>
<li><a href="https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/revert.sw"><code>revert</code></a>: A module containing
<ul>
<li><code>require</code>: A function that reverts and logs a given value if the condition is <code>false</code></li>
<li><code>revert</code>: A function that reverts</li>
</ul>
</li>
<li><a href="https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/logging.sw"><code>log</code></a>: A function that logs arbitrary stack types</li>
<li><a href="https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/auth.sw"><code>msg_sender</code></a>: A function that gets the Identity from which a call was made</li>
</ul>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="../../documentation/examples/liquidity-pool.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next prefetch" href="../../documentation/misc/language-comparison/index.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="../../documentation/examples/liquidity-pool.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>

                    <a rel="next prefetch" href="../../documentation/misc/language-comparison/index.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="../../elasticlunr.min.js"></script>
        <script src="../../mark.min.js"></script>
        <script src="../../searcher.js"></script>

        <script src="../../clipboard.min.js"></script>
        <script src="../../highlight.js"></script>
        <script src="../../book.js"></script>

        <!-- Custom JS scripts -->


    </div>
    </body>
</html>
