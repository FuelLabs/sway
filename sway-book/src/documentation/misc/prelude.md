# Standard Library Prelude

The [prelude](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/prelude.sw) is a list of commonly used features from the [standard library](https://github.com/FuelLabs/sway/tree/master/sway-lib-std) which is automatically imported into every Sway program.

The prelude contains the following modules:

- [`Address`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/address.sw): A struct containing a `b256` value which represents the wallet address
- [`assert`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/assert.sw): A function that reverts if the condition provided is `false`
- [`ContractId`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/contract_id.sw) A struct containing a `b256` value which represents the ID of a contract
- [`Identity`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/identity.sw): An enum containing `Address` & `ContractID` structs
- [`Option`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/option.sw): An enum containing either some generic value `<T>` or an absence of that value
- [`Result`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/result.sw): An enum used to represent either a success or failure of an operation
- [`revert`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/revert.sw): A module containing
  - `require`: A function that reverts and logs a given value if the condition is `false`
  - `revert`: A function that reverts
- [`StorageMap`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/storage.sw): A key-value mapping in contract storage
- [`Vec`](https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/vec.sw): A growable, heap-allocated vector
