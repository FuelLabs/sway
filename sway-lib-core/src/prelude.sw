library;

//! Defines the Sway core library prelude.
//! The prelude consists of implicitly available items,
//! for which `use` is not required.
pub use ::primitives::*;
pub use ::primitive_conversions::*;
pub use ::raw_ptr::*;
pub use ::raw_slice::*;
pub use ::never::*;
pub use ::ops::*;
pub use ::storage::*;
pub use ::str::*;
pub use ::codec::*;
