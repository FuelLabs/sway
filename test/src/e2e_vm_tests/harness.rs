use anyhow::{bail, Result};
use forc::test::{
    forc_abi_json, forc_build, forc_deploy, forc_run, BuildCommand, DeployCommand, JsonAbiCommand,
    RunCommand,
};
use fuel_tx::Transaction;
use fuel_vm::interpreter::Interpreter;
use fuel_vm::prelude::*;
use serde_json::Value;
use std::fs;

pub(crate) fn deploy_contract(file_name: &str) -> ContractId {
    // build the contract
    // deploy it
    println!(" Deploying {}", file_name);
    let manifest_dir = env!("CARGO_MANIFEST_DIR");

    let (verbose, use_orig_asm, use_orig_parser) = get_test_config_from_env();

    tokio::runtime::Runtime::new()
        .unwrap()
        .block_on(forc_deploy::deploy(DeployCommand {
            path: Some(format!(
                "{}/src/e2e_vm_tests/test_programs/{}",
                manifest_dir, file_name
            )),
            use_orig_asm,
            use_orig_parser,
            silent_mode: !verbose,
            ..Default::default()
        }))
        .unwrap()
}

/// Run a given project against a node. Assumes the node is running at localhost:4000.
pub(crate) fn runs_on_node(
    file_name: &str,
    contract_ids: &[fuel_tx::ContractId],
) -> Vec<fuel_tx::Receipt> {
    println!("Running on node: {}", file_name);
    let manifest_dir = env!("CARGO_MANIFEST_DIR");

    let mut contracts = Vec::<String>::with_capacity(contract_ids.len());
    for contract_id in contract_ids {
        let contract = format!("0x{:x}", contract_id);
        contracts.push(contract);
    }

    let (verbose, use_orig_asm, use_orig_parser) = get_test_config_from_env();

    let command = RunCommand {
        path: Some(format!(
            "{}/src/e2e_vm_tests/test_programs/{}",
            manifest_dir, file_name
        )),
        node_url: "http://127.0.0.1:4000".into(),
        use_orig_asm,
        use_orig_parser,
        silent_mode: !verbose,
        contract: Some(contracts),
        ..Default::default()
    };
    tokio::runtime::Runtime::new()
        .unwrap()
        .block_on(forc_run::run(command))
        .unwrap()
}

/// Very basic check that code does indeed run in the VM.
/// `true` if it does, `false` if not.
pub(crate) fn runs_in_vm(file_name: &str) -> ProgramState {
    let storage = MemoryStorage::default();

    let script = compile_to_bytes(file_name).unwrap();
    let gas_price = 10;
    let gas_limit = fuel_tx::consts::MAX_GAS_PER_TX;
    let byte_price = 0;
    let maturity = 0;
    let script_data = vec![];
    let inputs = vec![];
    let outputs = vec![];
    let witness = vec![];
    let tx_to_test = Transaction::script(
        gas_price,
        gas_limit,
        byte_price,
        maturity,
        script,
        script_data,
        inputs,
        outputs,
        witness,
    );
    let block_height = (u32::MAX >> 1) as u64;
    tx_to_test.validate(block_height).unwrap();
    let mut i = Interpreter::with_storage(storage);
    *i.transact(tx_to_test).unwrap().state()
}

/// Panics if code _does_ compile, used for test cases where the source
/// code should have been rejected by the compiler.
pub(crate) fn does_not_compile(file_name: &str) {
    assert!(
        compile_to_bytes(file_name).is_err(),
        "{} should not have compiled.",
        file_name,
    )
}

/// Returns `true` if a file compiled without any errors or warnings,
/// and `false` if it did not.
pub(crate) fn compile_to_bytes(file_name: &str) -> Result<Vec<u8>> {
    println!(" Compiling {}", file_name);
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    let (verbose, use_orig_asm, use_orig_parser) = get_test_config_from_env();
    forc_build::build(BuildCommand {
        path: Some(format!(
            "{}/src/e2e_vm_tests/test_programs/{}",
            manifest_dir, file_name
        )),
        use_orig_asm,
        use_orig_parser,
        silent_mode: !verbose,
        ..Default::default()
    })
    .map(|compiled| compiled.bytecode)
}

pub(crate) fn test_json_abi(file_name: &str) -> Result<()> {
    let _compiled_res = compile_to_json_abi(file_name)?;
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    let oracle_path = format!(
        "{}/src/e2e_vm_tests/test_programs/{}/{}",
        manifest_dir, file_name, "json_abi_oracle.json"
    );
    let output_path = format!(
        "{}/src/e2e_vm_tests/test_programs/{}/{}",
        manifest_dir, file_name, "json_abi_output.json"
    );
    if fs::metadata(oracle_path.clone()).is_err() {
        bail!("JSON ABI oracle file does not exist for this test.");
    }
    if fs::metadata(output_path.clone()).is_err() {
        bail!("JSON ABI output file does not exist for this test.");
    }
    let oracle_contents =
        fs::read_to_string(oracle_path).expect("Something went wrong reading the file.");
    let output_contents =
        fs::read_to_string(output_path).expect("Something went wrong reading the file.");
    if oracle_contents != output_contents {
        bail!("Mismatched ABI JSON output.");
    }
    Ok(())
}

fn compile_to_json_abi(file_name: &str) -> Result<Value> {
    println!("   ABI gen {}", file_name);
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    forc_abi_json::build(JsonAbiCommand {
        path: Some(format!(
            "{}/src/e2e_vm_tests/test_programs/{}",
            manifest_dir, file_name
        )),
        json_outfile: Some(format!(
            "{}/src/e2e_vm_tests/test_programs/{}/{}",
            manifest_dir, file_name, "json_abi_output.json"
        )),
        silent_mode: true,
        ..Default::default()
    })
}

fn get_test_config_from_env() -> (bool, bool, bool) {
    let var_exists = |key| std::env::var(key).map(|_| true).unwrap_or(false);

    (
        var_exists("SWAY_TEST_VERBOSE"),
        var_exists("SWAY_TEST_USE_ORIG_ASM"),
        var_exists("SWAY_TEST_USE_ORIG_PARSER"),
    )
}
