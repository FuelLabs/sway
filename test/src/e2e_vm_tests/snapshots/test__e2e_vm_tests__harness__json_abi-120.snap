---
source: test/src/e2e_vm_tests/harness.rs
expression: output_contents
---
[
  {
    "inputs": [],
    "name": "get_42",
    "outputs": [
      {
        "components": null,
        "name": "",
        "type": "u64",
        "typeArguments": null
      }
    ],
    "type": "function"
  }
]
